### PR TITLE
Make some changes to support `torch.compile` and add tests

### PR DESCRIPTION
**Pull request by danieldk**
## Description

Supporting `torch.compile` requires some small changes:

- We can't use `IntEnum`, replace single use by `Enum`.
- In the causal mask generation, the data type gets inferred incorrectly, explicitly give the `dtype`.

This change also adds tests for `torch.compile`. To avoid repetitiveness, factor out decoder tests similar to encoder tests.

In the `torch.compile` test for NeoX we disable testing with cache and positions. `torch.compile` is tripping over the rotary embeddings code where the rotary embeddings are only applied to part of the query/key representations.

While debugging, I also found that `GeluNew` can be implemented with fewer operations.

### Types of change

Feature/tests

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/layers/activations.py [hunk 1]
```diff
@@ -51,11 +51,7 @@ def forward(self, input: Tensor) -> Tensor:
...
-            * input
-            * (
-                1.0
-                + torch.tanh(input * 0.7978845608 * (1.0 + 0.044715 * input * input))
-            )
-        )
+        alpha = math.sqrt(2.0 / math.pi)
+        beta = 0.044715
+
+        return 0.5 * input * (1.0 + torch.tanh(alpha * (input + beta * input.pow(3))))
```
Comment (shadeMe) [inline curated_transformers/layers/activations.py:57]: Is there a reference for this?

File: curated_transformers/tests/models/gpt_neox/test_decoder.py [hunk 2]
```diff
@@ -1,103 +1,36 @@
...
-    torch.manual_seed(0)
-    X = torch.randint(0, hf_model.config.vocab_size, (2, 10), device=torch_device)
-    # NOTE: this test may break in the future because we are relying on a
-    #       slicing bug, which results in more general support for positions
-    #       (like our implementation):
-    #
-    #       https://github.com/huggingface/transformers/blob/f924df3c7e5150317ef47754a31cebf2893570ce/src/transformers/models/gpt_neox/modeling_gpt_neox.py#L278
+def test_decoder_with_torch_compile(torch_device):
+    # NOTE: testing with cache/positions fails. All outputs seems ok until
+    # the concatenation of the vectors with/without rotary embeddings:
```
Comment (shadeMe) [inline curated_transformers/tests/models/gpt_neox/test_decoder.py:26]: Shall we (temporarily) add an `xfail` test case for that?

File: curated_transformers/tests/models/util.py [hunk 3]
```diff
@@ -9,20 +9,152 @@
...
+
+    torch.manual_seed(0)
+    X = torch.randint(0, hf_model.config.vocab_size, (2, 10), device=torch_device)
+
+    with torch.no_grad():
+        Y = model(X).logits
+        Y_hf = hf_model(X).logits
+    torch_assertclose(Y, Y_hf, atol=atol, rtol=rtol)
+
+    mask = torch.rand((2, 10), dtype=torch.float, device=torch_device) < 0.5
```
Comment (shadeMe) [inline curated_transformers/tests/models/util.py:48]: Do we need to reset the manual seed for this invocation as well? I recall that we do that in the sampling generator tests.

File: curated_transformers/layers/activations.py (hunk 1 above)
Comment (danieldk) [reply curated_transformers/layers/activations.py:57]: No, I just rewrote the original equation a bit (multiplying out stuff).

File: curated_transformers/tests/models/util.py (hunk 3 above)
Comment (danieldk) [reply curated_transformers/tests/models/util.py:48]: I considered doing this, since it's a known failure. However, the problem is that if Torch fixes this issue, the test will suddenly pass (and thus fail), which is probably not nice.

File: curated_transformers/tests/models/gpt_neox/test_decoder.py (hunk 2 above)
Comment (danieldk) [reply curated_transformers/tests/models/gpt_neox/test_decoder.py:26]: I considered doing this, since it's a known failure. However, the problem is that if Torch fixes this issue, the test will suddenly pass (and thus fail), which is probably not nice.

File: curated_transformers/tests/models/util.py (hunk 3 above)
Comment (danieldk) [reply curated_transformers/tests/models/util.py:48]: That is so that we can reorder/remove particular generation tests in the future without failing the generation tests after that. But that's not an issue here, since we do not manually validate the output.
